File: go/0328_odd_even_linked_list.go
func oddEvenList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	if head.Next == nil {
		return head
	}

	oddHead, evenHead := head, head.Next
	oddCur, evenCur := oddHead, evenHead

	var oddStop, evenStop bool
	for !oddStop || !evenStop {
		if oddCur.Next != nil && oddCur.Next.Next != nil {
			oddCur.Next = oddCur.Next.Next
			oddCur = oddCur.Next
		} else {
			oddStop = true
		}
		if evenCur.Next != nil && evenCur.Next.Next != nil {
			evenCur.Next = evenCur.Next.Next
			evenCur = evenCur.Next
		} else {
			evenStop = true
		}
	}
	oddCur.Next, evenCur.Next = evenHead, nil
	return oddHead
}

File: python/0328_odd_even_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        dummy = head
        oddCur = head
        if not head or not head.next:
            return oddCur
        evenCur = head.next
        evenHead = evenCur

        while True:
            oddUpdate, evenUpdate = False, False

            if oddCur and oddCur.next:
                tmp = oddCur.next
                oddCur.next = oddCur.next.next
                if oddCur.next is not None:
                    oddUpdate = True
                    oddCur = oddCur.next
            if evenCur and evenCur.next:
                evenCur.next = evenCur.next.next
                if evenCur.next is not None:
                    evenUpdate = True
                    evenCur = evenCur.next

            if not oddUpdate and not evenUpdate:
                break

        oddCur.next = evenHead
        return dummy
